add support to unordered_set

nepluno opened the pull request:
This PR adds type castor for `std::unordered_set`, and is part of #84, which is split into two parts. Here we don't do the same for `std::set` since Python doesn't have an ordered set by default.

wjakob commented on include/nanobind/stl/detail/nb_set.h:
```diff
@@ -0,0 +1,56 @@
+        bool success = (size >= 0);
+
+        KeyCaster key_caster;
+        while (PyObject* key = obj_iter_next(iter)) {
```
potential issue: `obj_iter_next` can raise, but `from_python` cannot. We need a non-throwing version of that function which indicates a failure (and clears the underlying Python error status) so that the caster can return `false` in this case.

wjakob replied:
minor: is in't it common to put an extra pair of parentheses in a `while` statement when a variable is defined there? Perhaps it's cleaner to use a `for` statement (without upper bound) that has that idea builtin.

nepluno replied:
Thanks! I've changed this to a for-loop. And the error has been checked if `key` is a `nullptr` (exit the for-loop), where we will print & clear the error.

wjakob commented on include/nanobind/stl/detail/nb_set.h:
```diff
@@ -0,0 +1,56 @@
+            for (auto& key : src) {
+                object k = steal(
+                    KeyCaster::from_cpp(forward_like<T>(key), policy, cleanup));
+                if (PySet_Add(ret.ptr(), k.ptr()) != 0) return handle();
```
`PySet_Add` can fail with an error status set. In that case it's not valid to just return.

nepluno replied:
Thanks! I've added an error printing (and clearing by the way) there.

wjakob commented on include/nanobind/stl/detail/nb_set.h:
```diff
@@ -0,0 +1,56 @@
+
+    template <typename T>
+    static handle from_cpp(T &&src, rv_policy policy, cleanup_list *cleanup) {
+        object ret = steal(PySet_New(nullptr));
```
`PySet_New` can fail with an error status set. In that case it's not valid to just return.

wjakob commented on include/nanobind/stl/detail/nb_set.h:
```diff
@@ -0,0 +1,67 @@
+        KeyCaster key_caster;
+        for (PyObject* key = PyIter_Next(iter); key; key = PyIter_Next(iter)) {
+            if (!key_caster.from_python(key, flags, cleanup)) {
+                success = false;
```
missing DECREF in the case of a failure?

nepluno replied:
Added. Thanks!

wjakob commented on include/nanobind/stl/detail/nb_set.h:
```diff
@@ -0,0 +1,67 @@
+
+        PyObject* iter = obj_iter(src.ptr());
+        if (iter == nullptr) {
+            PyErr_Print();
```
`PyErr_Clear()`, not `PyErr_Print()`. The type casters should not print messages in the case of failure.

nepluno replied:
Replaced. Thanks.

wjakob commented on include/nanobind/stl/detail/nb_set.h:
```diff
@@ -0,0 +1,67 @@
+            return false;
+        }
+
+        Py_ssize_t size = NB_SET_GET_SIZE(src.ptr());
```
Looks to me like `size` isn't actually used ..

nepluno replied:
Removed. Thanks.

wjakob commented on include/nanobind/stl/detail/nb_set.h:
```diff
@@ -0,0 +1,67 @@
+                    KeyCaster::from_cpp(forward_like<T>(key), policy, cleanup));
+                if (PySet_Add(ret.ptr(), k.ptr()) != 0) {
+                    if (PyErr_Occurred()) {
+                        PyErr_Print();
```
ditto

nepluno replied:
> I left a few more comments

Thanks! I've fixed correspondingly.

wjakob commented on include/nanobind/stl/detail/nb_set.h:
```diff
@@ -0,0 +1,67 @@
+                if (!k.is_valid()) return handle();
+            }
+        } else if (PyErr_Occurred()) {
+            PyErr_Print();
```
ditto

wjakob commented on include/nanobind/stl/unordered_set.h:
```diff
@@ -0,0 +1,16 @@
+template <typename Key, typename Hash, typename Compare, typename Alloc>
+struct type_caster<std::unordered_set<Key, Hash, Compare, Alloc>>
+    : set_caster<std::unordered_set<Key, Hash, Compare, Alloc>, Key>
+{
```
brace on previous line

wjakob commented:
I left a few more comments

nepluno commented:
> I left a few more comments

Thanks! I've fixed correspondingly.

wjakob commented:
Hi @nepluno ,

I just took another look. is it possible that you forgot to stage/commit the python portion of the tests? I see C++ test functions `set_in_value`, `set_in_lvalue_ref`, and `set_in_rvalue_ref`, but those are never called from the Python test suite.

Finally, would you be able to add a test that corresponds to a type caster failure? For example, trying to pass a set containing a string to a function taking a `std::unordered_map<int>`. (Doesn't have to be exactly that, the important thing is just that we check that nanobind then refuses to run the associated function overload.)

Thanks,
Wenzel

nepluno commented:
> is it possible that you forgot to stage/commit the python portion of the tests?

Hi @wjakob! Sorry for the missing tests and thanks for the reminder! I've added them.

> would you be able to add a test that corresponds to a type caster failure?

Done. I've added a test using `unordered_set<int>` as the input to a function expecting `unordered_set<string>`.

wjakob commented:
Thanks!

wjakob commented:
I committed a few minor tweaks to the code in commit b26338fed6fc69d4a11795a610b460d82e711a26 and also added bindings for `std::set`. I don't think the different ordering convention is a big deal, and the same also applies to `std::map`. Python dict orders entries via the original insertion order, whereas the C++ one sorts by key. 🤷
